### ISSUE TITLE
[Support] TransferState still the proper way to path data between server and client?

### ISSUE THREAD

**Issue by peterpeterparker** [angular/universal-starter#414]
Hey,

As far as I understand, when I develop http queries, if I don't want them to be performed on both server and client side, I've to check the platform to only perform the queries on the server side and then pass the data to the client side thru the use of `TransferState`

Is this still the correct way to achieve http queries while using angular-universal/platform-server?

Thx in advance for the support

**Comment by peterpeterparker** [comment]
There isn't currently a "proper" way to do state transfer

A proposal / draft to handle this that should be implimented into angular/universal is a current work in progress

**Comment by denisyilmaz** [comment]
Im using a Route Resolver to check if the Data is already in Transferstate (then immediately resolve() the route), otherwise I start a HTTP Request for the data and add the data to Transferstate. 

This way you can assure that the data will be loaded only once without limiting it to the Server or Client side.

```typescript
import { Injectable, PLATFORM_ID, Inject } from '@angular/core';
import { isPlatformServer } from '@angular/common';
import { Resolve, RouterStateSnapshot, ActivatedRouteSnapshot } from '@angular/router';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Observable }     from 'rxjs/Observable';
import 'rxjs/Rx';

import { TransferState } from './../shared/modules/transfer-state/transfer-state';

@Injectable()
export class PageResolver implements Resolve<any> {
    constructor(
        private cache: TransferState,
        private http: HttpClient,
        @Inject(PLATFORM_ID) 
        private platformId: Object
    ) {}
    resolve(
        route: ActivatedRouteSnapshot,
        state: RouterStateSnapshot
    ): Observable<any>|Promise<any>|any {
        return new Promise((resolve, reject) => {
            let slug = route.params['slug'];
            let projectCacheSlug = 'page-' + slug;
            if(this.cache.get(projectCacheSlug)) {
                resolve()
            } else {
                let port = process.env.PORT || 8000
                let url = (isPlatformServer(this.platformId)) ? 'http://localhost:' + port : ''
                this.http
                    .get( url + '/api/pages/' + slug)
                    .retry(5)
                    .subscribe(
                        res   =>    {
                            this.cache.set(projectCacheSlug, res);
                            resolve();
                        },
                        (err: HttpErrorResponse) => {
                        if (err.error instanceof Error) {
                            // A client-side or network error occurred. Handle it accordingly.
                            console.log('An error occurred:', err.error.message);
                        } else {
                            // The backend returned an unsuccessful response code.
                            // The response body may contain clues as to what went wrong,
                            console.log(`Backend returned code ${err.status}, body was: ${err.error}`);
                        }
                        }
                    )
            }
        })
    }
}
```

Hope that helps.

**Comment by Toxicable** [comment]
StateTransfer has been merged into `@angular/platform-server`
https://github.com/angular/angular/commit/cfd9ca0d6f453decc154cc640351f5af67685510
It will be released in the next version

**Comment by MarkPieszak** [comment]
Yes for now you can use your own TransferState mechism, or wait until 5.0 to use the one built into Core. The webpack example here has an existing TransferState Http service you can use.
